csrc/scheduler/matmul_utils.cpp

```diff
@@ -398,6 +415,23 @@ std::shared_ptr<MatmulParams> getMatmulHeuristics(
   // Disable magic zero for matmul kernels
   params->cparams.enable_magic_zero = false;
 
+  // Disable shared memory epilogue before shared memory reuse is implemented.
+  //  Otherwise, there will be performance regression due to reduced occupancy
+  //  caused by extra shared memory usage.
+  constexpr bool allow_smem_epilogue = true;
```

Comment:
Is this still true with the new `hasEnoughSharedMemoryForEpilogue` that takes register occupancy into consideration?

Reply:
In case `FusionAmpereMatmulSmemEpilogue_CUDA`,  The settings of cta_tile (64, 128, 32), warp_tile(32, 32, 32), and smem_double_buffer_stage(2) are purposefully changed to produce a constant occupancy of 25% no matter `has_smem_epilogue = true or false`. This allows me to evaluate the influence of the has_smem_epilogue. There is no noticable change of performance for TT, NT, TN (kernel1-3). But for NN (kernel4), there is a slightly decrease. I checked the global memory write index, the access is coalesced for `has_smem_epilogue = true` and strided for `has_smem_epilogue = false` 

```
has_smem_epilogue = false
smem_false_64_128_32.log:kernel1 run in 1.19603 ms, achieved: 112.219 GB/s
smem_false_64_128_32.log:kernel2 run in 1.18682 ms, achieved: 113.091 GB/s
smem_false_64_128_32.log:kernel3 run in 1.00045 ms, achieved: 134.158 GB/s
smem_false_64_128_32.log:kernel4 run in 1.16224 ms, achieved: 115.482 GB/s

```

```
has_smem_epilogue = true
smem_true_64_128_32.log:kernel1 run in 1.19296 ms, achieved: 112.508 GB/s
smem_true_64_128_32.log:kernel2 run in 1.19091 ms, achieved: 112.702 GB/s
smem_true_64_128_32.log:kernel3 run in 0.995328 ms, achieved: 134.848 GB/s
smem_true_64_128_32.log:kernel4 run in 1.18067 ms, achieved: 113.679 GB/s
```
Global Memory write index for `MatmulLayout::NN with has_smem_epilogue = false`:
```
smem_false4_64_128_32_write_index
tidyz=00 tidx=0 from=0 to=0
tidyz=00 tidx=1 from=0 to=2
tidyz=00 tidx=2 from=0 to=4
tidyz=00 tidx=3 from=0 to=6
tidyz=00 tidx=4 from=0 to=4096
tidyz=00 tidx=5 from=0 to=4098
tidyz=00 tidx=6 from=0 to=4100
tidyz=00 tidx=7 from=0 to=4102
tidyz=00 tidx=8 from=0 to=8192
tidyz=00 tidx=9 from=0 to=8194
tidyz=00 tidx=10 from=0 to=8196
tidyz=00 tidx=11 from=0 to=8198
tidyz=00 tidx=12 from=0 to=12288
tidyz=00 tidx=13 from=0 to=12290
tidyz=00 tidx=14 from=0 to=12292
tidyz=00 tidx=15 from=0 to=12294
tidyz=00 tidx=16 from=0 to=16384
tidyz=00 tidx=17 from=0 to=16386
tidyz=00 tidx=18 from=0 to=16388
tidyz=00 tidx=19 from=0 to=16390
tidyz=00 tidx=20 from=0 to=20480
tidyz=00 tidx=21 from=0 to=20482
tidyz=00 tidx=22 from=0 to=20484
tidyz=00 tidx=23 from=0 to=20486
tidyz=00 tidx=24 from=0 to=24576
tidyz=00 tidx=25 from=0 to=24578
tidyz=00 tidx=26 from=0 to=24580
tidyz=00 tidx=27 from=0 to=24582
tidyz=00 tidx=28 from=0 to=28672
tidyz=00 tidx=29 from=0 to=28674
tidyz=00 tidx=30 from=0 to=28676
tidyz=00 tidx=31 from=0 to=28678

```

Global Memory write index for `MatmulLayout::NN with has_smem_epilogue = true`:
```
smem_true4_64_128_32_write_index
tidyz=00 tidx=0 from=0 to=0
tidyz=00 tidx=1 from=4 to=4
tidyz=00 tidx=2 from=8 to=8
tidyz=00 tidx=3 from=12 to=12
tidyz=00 tidx=4 from=16 to=16
tidyz=00 tidx=5 from=20 to=20
tidyz=00 tidx=6 from=24 to=24
tidyz=00 tidx=7 from=28 to=28
tidyz=00 tidx=8 from=32 to=32
tidyz=00 tidx=9 from=36 to=36
tidyz=00 tidx=10 from=40 to=40
tidyz=00 tidx=11 from=44 to=44
tidyz=00 tidx=12 from=48 to=48
tidyz=00 tidx=13 from=52 to=52
tidyz=00 tidx=14 from=56 to=56
tidyz=00 tidx=15 from=60 to=60
tidyz=00 tidx=16 from=64 to=64
tidyz=00 tidx=17 from=68 to=68
tidyz=00 tidx=18 from=72 to=72
tidyz=00 tidx=19 from=76 to=76
tidyz=00 tidx=20 from=80 to=80
tidyz=00 tidx=21 from=84 to=84
tidyz=00 tidx=22 from=88 to=88
tidyz=00 tidx=23 from=92 to=92
tidyz=00 tidx=24 from=96 to=96
tidyz=00 tidx=25 from=100 to=100
tidyz=00 tidx=26 from=104 to=104
tidyz=00 tidx=27 from=108 to=108
tidyz=00 tidx=28 from=112 to=112
tidyz=00 tidx=29 from=116 to=116
tidyz=00 tidx=30 from=120 to=120
tidyz=00 tidx=31 from=124 to=124


```

Reply:
so changing `has_smem_epilogue` to true makes the memory access contiguous, but the perf drop? That is strange.